tensorizer/serialization.py

```diff
@@ -2432,384 +2192,283 @@ def _generate_state_dict(self) -> None:
+                # if this tensor would make this current reader more than ideal_tensor_bytes_per_reader, and there's at least one existing item in the reader
+                if running_total + tensor_sizes[tensor_size_idx][1] > ideal_tensor_bytes_per_reader and chunk_start != tensor_size_idx:
+                    # break it
+                    tensors_per_reader.append(tensor_sizes[chunk_start:tensor_size_idx])
+                    chunk_start = tensor_size_idx
+                    running_total = 0
+                running_total += tensor_sizes[tensor_size_idx][1]
+
+            if running_total:
+                # The last one gets the leftovers
+                if not tensors_per_reader:
+                    tensors_per_reader.append([])
+
+                tensors_per_reader[-1].extend(tensor_sizes[chunk_start:])
+
+        transfer_out_queue = queue.SimpleQueue()  # type: queue.SimpleQueue[Union[Exception, TensorDeserializer._CopiedData]]
+
+        copy_threads = []
+        barrier = threading.Barrier(len(tensors_per_reader))
+        halt = AtomicUint(width=4)
+
+        for thread_idx, tensor_items in enumerate(tensors_per_reader):
+            # print(f'Thread #{thread_idx} will read {len(tensor_items)} tensors, {sum(v for _, v in tensor_items)/1024/1024:.3f}MiB, {max(v for _, v in tensor_items)/1024/1024:.3f}MiB max', file=sys.stderr)
+
+            thread = threading.Thread(
+                name=f'TensorDeserializerCopy-{thread_idx}',
+                target=self._copy_thread,
+                args=(
+                    thread_idx,
+                    halt,
+                    barrier,
+                    verify_hash,
+                    tensor_items,
+                    transfer_out_queue)
+            )
+            copy_threads.append(thread)
+            thread.start()
+
+        for _ in range(len(keys)):
+            copied_data = transfer_out_queue.get(timeout=3600)
+            if isinstance(copied_data, Exception):
+                halt.store(1)
+                # error occurred; halt
+                raise copied_data
+
+            name = copied_data.header.name
+            self._cache[name] = copied_data
+            yield copied_data
+
+    def _copy_thread(
+            unsafe_self,
+            thread_idx: int,
+            halt: AtomicUint,
+            barrier: threading.Barrier,
+            verify_hash: bool,
+            tensor_items: Sequence[Tuple[str, int]],  # (name, size)
+            transfer_out_queue  # type: queue.SimpleQueue[Union[Exception, TensorDeserializer._CopiedData]]
+    ):
+        # Need to get rid of self or more safely have thread-local storage
 
-        def _is_done(self):
-            return self._count == 0 or self._cancelled
+        # cuda stream. First one is slow, spin it off? This is gross TOOD bchess
+        is_cuda = unsafe_self._device.type == 'cuda'
+        cuda_stream = None
+        if is_cuda:
+            cuda_stream = torch.cuda.Stream()
```

Comment:
CUDA streams are tied to a particular device, so `unsafe_self._device` should be chosen in case tensors are being deserialized to a specific non-default device
```suggestion
            cuda_stream = torch.cuda.Stream(unsafe_self._device)
```